cedar-drt/src/lean_impl.rs

```diff
@@ -0,0 +1,249 @@
+pub use lean_sys::lean_object;
+pub use lean_sys::string::lean_mk_string;
+use lean_sys::{
+    lean_initialize_runtime_module, lean_io_mark_end_initialization, lean_io_mk_world,
+    lean_string_cstr,
+};
+use serde::{Deserialize, Serialize};
+use std::ffi::CStr;
+use std::str::FromStr;
+
+#[link(name = "Cedar", kind = "static")]
+#[link(name = "Std", kind = "static")]
+#[link(name = "DiffTest", kind = "static")]
+extern "C" {
+    fn isAuthorizedDRT(req: *mut lean_object) -> *mut lean_object;
+    fn validateDRT(req: *mut lean_object) -> *mut lean_object;
+    fn initialize_DiffTest_Main(builtin: i8, ob: *mut lean_object) -> *mut lean_object;
+}
+
+#[derive(Serialize, Deserialize)]
+struct ListDef<String> {
+    l: Vec<String>,
+}
+
+#[derive(Serialize, Deserialize)]
+struct SetDef<String> {
+    mk: ListDef<String>,
+}
+
+#[derive(Serialize, Deserialize)]
+struct ResponseDef {
+    policies: SetDef<String>,
+    decision: String,
+}
+
+#[derive(Serialize, Deserialize)]
+enum ValResponseDef {
+    #[serde(rename = "ok")]
+    Ok,
+    #[serde(rename = "error")]
+    Error(String),
+}
+
+pub struct LeanDefinitionalEngine {}
+
+fn lean_obj_to_string(o: *mut lean_object) -> String {
+    let lean_obj_p = unsafe { lean_string_cstr(o) };
+    let lean_obj_cstr = unsafe { CStr::from_ptr(lean_obj_p as *const i8) };
+    lean_obj_cstr.to_string_lossy().into_owned() //TODO: lossy
+}
+
+impl LeanDefinitionalEngine {
+    pub fn new() -> Self {
+        if env::var("RUST_LEAN_INTERFACE_INIT").is_err() {
+            unsafe { lean_initialize_runtime_module() };
+            unsafe { lean_initialize() };
+            unsafe { initialize_DiffTest_Main(1, lean_io_mk_world()) };
+            unsafe { lean_io_mark_end_initialization() };
+            env::set_var("RUST_LEAN_INTERFACE_INIT", "1");
+        }
+        Self {}
+    }
+
+    fn serialize_authorization_request(
+        request: &ast::Request,
+        policies: &ast::PolicySet,
+        entities: &Entities,
+    ) -> *mut lean_object {
+        let request: String = serde_json::to_string(&RequestForDefEngine {
+            request,
+            policies,
+            entities,
+        })
+        .expect("Failed to serialize request, policies, or entities");
+        let cstring = CString::new(request).expect("CString::new failed");
+        let s = unsafe { lean_mk_string(cstring.as_ptr() as *const u8) };
+        return s;
+    }
+
+    fn deserialize_authorization_response(response: *mut lean_object) -> InterfaceResponse {
+        let response_string = lean_obj_to_string(response);
+        let resp: ResponseDef =
+            serde_json::from_str(&response_string).expect("could not deserialize json");
+        let dec: authorizer::Decision = if resp.decision == "allow" {
+            authorizer::Decision::Allow
+        } else if resp.decision == "deny" {
+            authorizer::Decision::Deny
+        } else {
+            panic!("unknown decision")
+        };
```

Comment:
We can't reuse `authorizer::Decision` directly because it serializes as "Allow"/"Deny" while the Lean serializes as "allow"/"deny". I'd rather leave both definitions as-is to match existing coding style, but I'll switch to a match statement in my next commit to make this a little prettier.